src/rust/engine/options/src/args.rs

```diff
@@ -3,95 +3,142 @@
 
 use std::env;
 
-use super::id::{NameTransform, OptionId, Scope};
+use super::id::{is_valid_scope_name, NameTransform, OptionId, Scope};
 use super::{DictEdit, OptionsSource};
-use crate::parse::{expand, expand_to_dict, expand_to_list, Parseable};
+use crate::parse::{expand, expand_to_dict, expand_to_list, ParseError, Parseable};
 use crate::ListEdit;
-use std::collections::HashMap;
-
-pub struct Args {
-    pub(crate) args: Vec<String>,
+use core::iter::once;
+use itertools::{chain, Itertools};
+
+#[derive(Debug)]
+struct Arg {
+    context: Scope,
+    flag: String,
+    value: Option<String>,
 }
 
-#[derive(PartialEq)]
-enum Negate {
-    True,
-    False,
-}
+impl Arg {
+    /// Checks if this arg's flag is equal to the provided strings concatenated with dashes.
+    /// E.g., "--foo-bar" matches ["-", "foo", "bar"].
+    fn flag_match<'a>(&self, dash_separated_strs: impl Iterator<Item = &'a str>) -> bool {
+        #[allow(unstable_name_collisions)]
+        // intersperse is provided by itertools::Itertools, but is also in the Rust nightly
+        // as an experimental feature of standard Iterator. If/when that becomes standard we
+        // can use it, but for now we must squelch the name collision.
+        itertools::equal(
+            self.flag.chars(),
+            dash_separated_strs
+                .map(str::chars)
+                .intersperse("-".chars())
+                .flatten(),
+        )
+    }
 
-impl Args {
-    pub fn new(args: Vec<String>) -> Self {
-        Self { args }
+    /// Checks if this arg provides a value for the specified option, either negated or not.
+    fn _matches(&self, id: &OptionId, negate: bool) -> bool {
+        fn prefix<'a>(negate: bool) -> impl Iterator<Item = &'a str> {
+            if negate {
+                once("--no")
+            } else {
+                once("-")
+            }
+        }
+        // Check if --scope-flag matches,
+        self.flag_match(chain![
+            prefix(negate),
+            once(id.scope.name()),
+            id.name_components_strs()
+        ]) ||
+        // Check if --flag matches in the context of the current goal's scope.
+        (self.context == id.scope &&
+         self.flag_match(chain![prefix(negate), id.name_components_strs()])) ||
+        // Check if -s matches for a short name s, if any.
+        (if let Some(sn) = &id.short_name {
+            self.flag_match(chain![once(""), once(sn.as_ref())])
```

Comment:
What does `once("")` mean?

Reply:
It's a rust-ism for "an iterator that yields this argument once"

Reply:
Why is it necessary to chain a leading empty string here?

Reply:
To get the single dash following it by `intersperse`.

Reply:
Ah, okay. This code is fine, but, for future, I wonder if something like `["", sn.as_ref()].into_iter()` might work too.

Reply:
If that seems more readable to you I can do that.

Reply:
On second thoughts, since we must use `chain!` and `once()` elsewhere in this file, it makes sense to do so here, for uniformity.